src/protocol/attribution/accumulate.rs

```diff
@@ -0,0 +1,322 @@
+                    ctx,
+                    step_size,
+                    current,
+                    successor,
+                    RecordId::from(i),
+                    [steps[1], steps[2], steps[3], steps[4]],
+                ));
+            }
+
+            let results = try_join_all(accumulation_futures).await?;
+
+            // save the calculation results to these support vectors for use in the next iteration
+            results
+                .into_iter()
+                .enumerate()
+                .for_each(|(i, (credit, stop_bit))| {
+                    credits[i] += credit;
+                    stop_bits[i] = stop_bit;
+                });
+        }
+
+        // drop irrelevant fields and add another supporting field called `aggregation_bit` for the next step
+        let output: Batch<AccumulationOutputRow<F>> = self
+            .input
+            .iter()
+            .enumerate()
+            .map(|(i, x)| AccumulationOutputRow {
+                breakdown_key: x.breakdown_key,
+                credit: credits[i],
+                aggregation_bit: one,
+            })
+            .collect::<Vec<_>>()
+            .into();
+
+        // TODO: Append unique breakdown_key values at the end of the output vector for the next step
+        // Since we cannot see the actual breakdown key values, we'll append shares of [0..MAX]. Adding u8::MAX
+        // number of elements to the output won't be a problem. Adding u32::MAX elements will be 1B + u32::MAX, which
+        // exceeds our current assumption of `input.len() < 1B`.
+
+        Ok(output)
+    }
+
+    async fn get_accumulated_credit<S, N>(
+        &self,
+        ctx: &ProtocolContext<'_, S, N>,
+        step_size: u32,
+        current: AccumulationInputRow<F>,
+        successor: AccumulationInputRow<F>,
+        record_id: RecordId,
+        steps: [S; 4],
+    ) -> Result<(Replicated<F>, Replicated<F>), BoxError>
+    where
+        S: Step + SpaceIndex,
+        N: Network<S>,
+    {
+        // first, calculate [successor.helper_bit * successor.trigger_bit]
+        let mut b = ctx
+            .multiply(record_id, steps[0])
+            .await
+            .execute(successor.report.helper_bit, successor.report.is_trigger_bit)
+            .await?;
+
+        // since `stop_bits` is initialized with `[1]`s, we only multiply `stop_bit` in the second and later iterations
+        if step_size > 1 {
+            b = ctx
+                .multiply(record_id, steps[1])
+                .await
+                .execute(b, current.stop_bit)
+                .await?;
+        }
+
+        let credit_future = ctx
+            .multiply(record_id, steps[2])
+            .await
+            .execute(b, successor.credit);
+
+        // for the same reason as calculating [b], we skip the multiplication in the first iteration
+        let stop_bit_future = if step_size > 1 {
+            futures::future::Either::Left(
```

Comment:
What does this do?

Reply:
`Either::Left` and `Either::Right` let you return futures conditionally. For this case, I want to skip the multiplication in the first iteration, so I return a future that resolves to `b` immediately when awaited.